pkg/apis/batch/validation/validation.go

```diff
@@ -389,8 +393,9 @@ func validatePodFailurePolicyRuleOnExitCodes(onExitCode *batch.PodFailurePolicyO
 }
 
 // validateJobStatus validates a JobStatus and returns an ErrorList with any errors.
-func validateJobStatus(status *batch.JobStatus, fldPath *field.Path) field.ErrorList {
+func validateJobStatus(job *batch.Job, fldPath *field.Path) field.ErrorList {
```

Comment:
In general, when strengthening validation, we should allow violations if already present.

The code looks similar to this https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#new-enum-value-in-existing-field, but in reverse:
If the old object doesn't pass the new validations, then we allow the update to also not pass them.

Reply:
But in this case this is status field validation (not spec), and the API object is under built-in controller only, so I think it falls under this paragraph from https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#backward-compatibility-gotchas: 
"Status fields whose writers are under our control (e.g., written by non-pluggable controllers), may potentially tighten validation, since that would cause a subset of previously valid values to be observable by clients."

It is only from now on, that tightening status validation should be respecting invalid old values. 

Please let me know if I'm missing something.

Reply:
While we expect to control status and we don't expect any difficulty, it's better to ratchet validation to account for the "what if we're wrong" case.  We'd be very disappointed if messed up conditions caused status updates to other fields to start failing.  It's a little more work, but it's time well spent.

Reply:
Sgtm, in that case I will move all the validation checks to the update. And only check them if the oldJob passes them. 

Still, we didn't support external controllers, then invalid field would mean a bug in the built controller or external agent mutating the status. If this is the case of a bug, then it is possible that the old job status is ok, but the controller wants to do update that violates the validation and fails. Same for external agents actually. I guess this is acceptable, right?

Reply:
Both points seem acceptable to me.

Reply:
Ok, reading this again:

>  We'd be very disappointed if messed up conditions caused status updates to other fields to start failing. 

I just check the validation if the related fields changed. This will allow status updates to the unrelated fields.

Alternatively, I could validate only the previous state was passing validation (as I initially thought), but maybe this is not needed. 

In both cases I don't fail status updates to unrelated fields, but in the proposed approach I fail when a user tries to change one invalid value to another invalid.